contracts/examples/WitnetRNG.sol

```diff
@@ -0,0 +1,338 @@
+            ? _latest
+            // start search from the latest block
+            : _searchPrevBlock(_block, __randomize_[_latest].prevBlock)
+        );
+    }
+
+    /// Returns `true` only when the randomness request that got posted within given block was already
+    /// reported back from the Witnet oracle, either successfully or with an error of any kind.
+    function isRandomized(uint256 _block)
+        public view
+        virtual override
+        returns (bool)
+    {
+        RandomizeData storage _data = __randomize_[_block];
+        return (
+            _data.witnetQueryId != 0 
+                && _witnetCheckResultAvailability(_data.witnetQueryId)
+        );
+    }
+
+    /// Generates a pseudo-random number uniformly distributed within the range [0 .. _range), by using 
+    /// the given `_nonce` value and the randomness returned by `getRandomnessAfter(_block)`. 
+    /// @dev Fails under same conditions as `getRandomnessAfter(uint256)` may do.
+    /// @param _range Range within which the uniformly-distributed random number will be generated.
+    /// @param _nonce Nonce value enabling multiple random numbers from the same randomness value.
+    /// @param _block Block number from which the search will start.
+    function random(uint32 _range, uint256 _nonce, uint256 _block)
+        external view
+        virtual override
+        returns (uint32)
+    {
+        return random(
+            _range,
+            _nonce,
+            keccak256(
+                abi.encode(
+                    msg.sender,
+                    getRandomnessAfter(_block)
+                )
+            )
+        );
+    }
+
+    /// Generates a pseudo-random number uniformly distributed within the range [0 .. _range), by using 
+    /// the given `_nonce` value and the given `_seed` as a source of entropy.
+    /// @param _range Range within which the uniformly-distributed random number will be generated.
+    /// @param _nonce Nonce value enabling multiple random numbers from the same randomness value.
+    /// @param _seed Seed value used as entropy source.
+    function random(uint32 _range, uint256 _nonce, bytes32 _seed)
+        public pure
+        virtual override
+        returns (uint32)
+    {
+        uint8 _flagBits = uint8(255 - _msbDeBruijn32(_range));
+        uint256 _number = uint256(
+                keccak256(
+                    abi.encode(_seed, _nonce)
+                )
+            ) & uint256(2 ** _flagBits - 1);
+        return uint32((_number * _range) >> _flagBits);
+    }
+
+    /// Requests the Witnet oracle to generate an EVM-agnostic and trustless source of randomness. 
+    /// Only one randomness request per block will be actually posted to the WRB. Should there 
+    /// already be a posted request within current block, all received funds shall be transfered
+    /// back to the tx sender.
+    function randomize()
+        external payable
+        virtual override
+        returns (uint256 _id)
```

Comment:
It says that this function returns an `_id`, but the documentation does not explain what is this `_id` and what to do with it.

I expected this to be a "randomness query id", such that the client will need this id to get their random number later. However that's not the case, the client must use a "uint256 block" in order to get the randomness. I assume that's the ethereum block number but it's also not clear by reading the documentation. Should that be the same block number as the one from the call to randomize? So I'm missing some documentation on how to use this contract.

Maybe it will all be explained in a high level tutorial somewhere else, but by reading the documentation of this file, if I were creating a solidity app I wouldn't know how to integrate this contract safely.

Reply:
Indeed, returning the queryId as it doing makes no sense, as there are no methods in the interface accepting this queryId as a parameter. It could return block.number instead, but this is something the calling contract already knows so there's no actual need either to return this value. So yep, it seems better not to return any value at all from randomize().